deepchem/data/data_loader.py

```diff
@@ -401,6 +401,136 @@ def _featurize_shard(self,
+    subsequently used for prediction by models. Upon prediction
+    atoms responsible for the activity modelled can be detected.
+
+    `FragmentLoader` is an abstract subclass of `DataLoader`. This class should
+    never be instantiated directly.  To load  dataset of fragments, use
+    subclasses (CSVFragmentLoader, SDFFragmentLoader) with
+    `ConvMolFeaturizer(per_atom_fragmentation=True)`. Then
+    call the `create_dataset()` method on (a list of) input
+    file(s) that hold the source data. Under the hood molecules
+    will be  fragmented, so that each fragment will represent
+    an atom-depleted version of  parent molecule (repeat for each atom). Fragments
+    featurized will be returned.
+
+    References
+    ---------
+
+    .. [1] Polishchuk, P., et al. J. Chem. Inf. Model. 2016, 56, 8, 1455–1469
+
+    Note
+    _________
+    Detailed examples of `GraphConvModel` interpretation are provided in Tutorial #28
+   """
+  def create_dataset(self,
+                     inputs: OneOrMany[Any],
+                     data_dir: Optional[str] = None,
+                     shard_size: Optional[int] = 8192) -> Dataset:
+    """Overrides `DataLoader`'s  `create_dataset()` method.
+    The only difference from parent is that it "parses"
+    fragmented molecules and assigns ids to fragments.
+    Creates and returns a `Dataset` object by featurizing provided files.
+
+    Reads in `inputs` and uses `self.featurizer` to featurize the
+    data in these inputs.  For large files, automatically shards
+    into smaller chunks of `shard_size` datapoints for convenience.
+    Returns a `Dataset` object that contains the featurized dataset.
+
+    This implementation assumes that the helper methods `_get_shards`
+    and `_featurize_shard` are implemented and that each shard
+    returned by `_get_shards` is a pandas dataframe.  You may choose
+    to reuse or override this method in your subclass implementations.
+
+    Parameters
+    ----------
+    inputs: List
+      List of inputs to process. Entries can be filenames or arbitrary objects.
+    data_dir: str, optional (default None)
+      Directory to store featurized dataset.
+    shard_size: int, optional (default 8192)
+      Number of examples stored in each shard.
+
+    Returns
+    -------
+    DiskDataset
+      A `DiskDataset` object containing a featurized representation of data
+      from `inputs`.
+    """
+    logger.info("Loading raw samples now.")
+    logger.info("shard_size: %s" % str(shard_size))
+
+    # Special case handling of single input
+    if not isinstance(inputs, list):
+      inputs = [inputs]
+
+    def shard_generator():
+      for shard_num, shard in enumerate(self._get_shards(inputs, shard_size)):
+        time1 = time.time()
+        X, valid_inds = self._featurize_shard(shard)
+        ids = shard[self.id_field].values
+        ids = ids[valid_inds]
+        ids = np.repeat(
```

Comment:
If I'm understanding correctly, the major change here is basically that we want all fragments that come from the same compound to have the same "id"?

It feels there should be a better abstraction here. Typically, we make a new `DataLoader` subclass for a new filetype. So it wouldn't quite make sense for us to have a `FragmentLoader` since there is no fragment filetype.

As one suggestion, why not just change the featurizer so that the feature for a molecule is of type `List[ConvMol]` rather than `ConvMol`? That is, for fragments, `ConvMolFeaturizer` returns a list of `ConvMol` rather than a single `ConvMol`. That means we don't have to change the basic abstraction of `DataLoader` for this one special case. We can also remove `CSVFragmentLoader` and `SDFFragmentLoader` as well

Reply:
> As one suggestion, why not just change the featurizer so that the feature for a molecule is of type `List[ConvMol]` rather than `ConvMol`? That is, for fragments, `ConvMolFeaturizer` returns a list of `ConvMol` rather than a single `ConvMol`. That means we don't have to change the basic abstraction of `DataLoader` for this one special case. We can also remove `CSVFragmentLoader` and `SDFFragmentLoader` as well

I prepred a new PR